### DISCUSSION TITLE
IndexError: list index out of range when using a transformer in pipeline

### DISCUSSION THREAD

**Discussion by oadams** [explosion/spaCy#8154] [Help: Coding & Implementations] [answered]
## How to reproduce the behaviour

I want to include a transformer in a spacy pipeline. Using the below code I get an `IndexError: list index out of range` in `spacy_transformers/layers/transformer_model.py", line 121`

```
from spacy_transformers import Transformer, TransformerModel
from spacy.lang.en import English

nlp = English()

# Construction via add_pipe with custom config
config = {
        "model": {
                    "@architectures": "spacy-transformers.TransformerModel.v1",
                    "name": "bert-base-uncased",
                    "tokenizer_config": {"use_fast": True}
                }
}
nlp.add_pipe("transformer", config=config)

doc = nlp('Here is some text')
```

If I replace `nlp = English()` with `nlp = spacy.load("en_core_web_trf")` and run `nlp.remove_pipe('transformer')` before adding the new transformer component, I still get the same error.

If I just use the transformer `nlp = spacy.load("en_core_web_trf")`. I don't get an error when processing docs, but I want to use my own custom transformer in the pipeline.

Note that I'm essentially following the code example for adding pipes at [https://spacy.io/api/transformer#init](https://spacy.io/api/transformer#init). 

## Your Environment

* Operating System: Tried on Mac and Linux
* Python Version Used: Python 3.8
* spaCy Version Used: spacy==3.0.6, spacy-alignments==0.8.3, spacy-legacy==3.0.5, spacy-transformers==1.0.2
* Environment Information:

**Comment by adrianeboyd** [comment, marked as answer]
Hi, this is a change in v3.0, most components require an additional `initialize` step to be ready for training/use:

```python
nlp.add_pipe("transformer", config=config)
# initialize the whole pipeline
nlp.initialize()
```

Or just the transformer component individually (if you're adding it an existing pipeline, running `nlp.initialize()` will wipe out all the weights of the **other** components):

```python
trf = nlp.add_pipe("transformer", config=config)
# initialize just the transformer
trf.initialize(lambda: iter([]), nlp=nlp)
```

See also: #7954
